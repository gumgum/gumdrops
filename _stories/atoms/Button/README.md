## Props

The following props may be passed to configure the button. All additional props are passed through to the `<button>` element:

| name      | type      | description                                                                     | default   |
| --------- | --------- | ------------------------------------------------------------------------------- | --------- |
| className | `String`  | Additional class to be added to the outermost element                           |           |
| isBlock   | `Boolean` | Indicate if this is a block element button                                      | `false`   |
| isGroup   | `Boolean` | Indicate if this part of a button group                                         | `false`   |
| size      | `String`  | Indicate the size of the button. One of `xs`, `sm` or `lg`                      |           |
| context   | `String`  | Indicate the context of the button. Example `default`, `warning`, `danger` etc. | `default` |
