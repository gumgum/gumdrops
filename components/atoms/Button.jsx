import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

const Button = ({
    context,
    type,
    size,
    isGroup,
    group,
    isBlock,
    onClick,
    className,
    style,
    children,
    ...otherProps
}) => {
    const baseClass = 'gds-button';

    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${context}`]: context,
        [`${baseClass}--${size}`]: size,
        [`${baseClass}-group__button`]: isGroup || group,
        [`${baseClass}--block`]: isBlock
    });

    return (
        <button className={rootClass} type={type} style={style} onClick={onClick} {...otherProps}>
            {children}
        </button>
    );
};

Button.displayName = 'Button';

Button.defaultProps = {
    context: 'default',
    type: 'button',
    isGroup: false,
    group: false,
    isBlock: false,
    onClick: null,
    style: {}
};

Button.propTypes = {
    /** outline, primary, secondary, success, warning, info, danger */
    context: PropTypes.string,
    type: PropTypes.string,
    /** xs, sm, lg */
    size: PropTypes.oneOf(['xs', 'sm', 'lg']),
    isGroup: PropTypes.bool,
    group: PropTypes.bool,
    isBlock: PropTypes.bool,
    onClick: PropTypes.func,
    className: PropTypes.string,
    style: PropTypes.object,
    children: PropTypes.node
};

export default Button;
