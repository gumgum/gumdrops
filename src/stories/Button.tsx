import React from 'react';
import cx from 'classnames';
import { Colors, Sizes } from 'types';

export interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement> {
    color?: Colors;
    type?: 'button' | 'submit';
    size?: Sizes;
    isGroup?: boolean;
    isBlock?: boolean;
    disabled?: boolean;
}

const baseClass = 'gds-button';

export const Button: React.FC<ButtonProps> = ({
    color = 'default',
    size = 'md',
    type = 'button',
    isGroup,
    isBlock,
    children,
    disabled,
    className,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${color}`]: color,
        [`${baseClass}--${size}`]: size,
        [`${baseClass}-group__button`]: isGroup,
        [`${baseClass}--block`]: isBlock,
        '-disabled': disabled,
        '-pointer-events--none': disabled
    });

    return (
        <button className={rootClass} type={type} disabled={disabled} {...otherProps}>
            {children}
        </button>
    );
};
