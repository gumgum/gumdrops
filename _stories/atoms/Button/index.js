import React from 'react';
import { text, select, boolean, object } from '@storybook/addon-knobs';
import { optionalSelect } from '../../../components/utils/optionalSelect';
import { action } from '@storybook/addon-actions';

import readme from './README.md';
import Button from '../../../components/atoms/Button';

const sizeOptions = {
    xs: 'xs',
    sm: 'sm',
    lg: 'lg',
    'No Value': ''
};

const contextOptions = {
    default: 'default',
    outline: 'outline',
    primary: 'primary',
    secondary: 'secondary',
    success: 'success',
    warning: 'warning',
    info: 'info',
    danger: 'danger'
};

const component = () => (
    <Button
        type={text('Type', 'button')}
        size={optionalSelect('Size', sizeOptions, '')}
        context={select('Context', contextOptions, 'default')}
        isGroup={boolean('isGroup', false)}
        isBlock={boolean('isBlock', false)}
        onClick={action('button_clicked')}
        className={text('ClassName', '')}
        style={object('Style', {})}>
        <i className="fa fa-check -m-r-2" />
        Great Success
    </Button>
);

export default [readme, component];
