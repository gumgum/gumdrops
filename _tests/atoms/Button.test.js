/* globals mount */
import React from 'react';
import Button from '../../components/atoms/Button';

const defaultProps = {
    isGroup: false,
    isBlock: false,
    size: 'sm',
    onClick: () => {},
    text: 'This is text',
    context: 'danger',
    className: 'foo-class',
    style: {
        width: '1000px'
    },
    onBlur: () => {}
};

describe('Expect <Button>', () => {
    it('to render', () => {
        const wrapper = mount(<Button {...defaultProps}>My Cool Button</Button>);
        expect(wrapper).toMatchSnapshot();
    });

    it('to renders as a group button', () => {
        const props = {
            ...defaultProps,
            isGroup: true
        };
        const wrapper = mount(<Button {...props}>My Cool Button</Button>);
        expect(wrapper).toMatchSnapshot();
    });

    it('to renders as a blocked button', () => {
        const props = {
            ...defaultProps,
            isBlock: true
        };
        const wrapper = mount(<Button {...props}>My Cool Button</Button>);
        expect(wrapper).toMatchSnapshot();
    });
});
