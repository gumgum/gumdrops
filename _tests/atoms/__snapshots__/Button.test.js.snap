// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Button> to render 1`] = `
<Button
  className="foo-class"
  context="danger"
  isBlock={false}
  isGroup={false}
  onBlur={[Function]}
  onClick={[Function]}
  size="sm"
  style={
    Object {
      "width": "1000px",
    }
  }
  text="This is text"
  type="button"
>
  <button
    className="gds-button foo-class gds-button--danger gds-button--sm"
    onBlur={[Function]}
    onClick={[Function]}
    style={
      Object {
        "width": "1000px",
      }
    }
    text="This is text"
    type="button"
  >
    My Cool Button
  </button>
</Button>
`;

exports[`Expect <Button> to renders as a blocked button 1`] = `
<Button
  className="foo-class"
  context="danger"
  isBlock={true}
  isGroup={false}
  onBlur={[Function]}
  onClick={[Function]}
  size="sm"
  style={
    Object {
      "width": "1000px",
    }
  }
  text="This is text"
  type="button"
>
  <button
    className="gds-button foo-class gds-button--danger gds-button--sm gds-button--block"
    onBlur={[Function]}
    onClick={[Function]}
    style={
      Object {
        "width": "1000px",
      }
    }
    text="This is text"
    type="button"
  >
    My Cool Button
  </button>
</Button>
`;

exports[`Expect <Button> to renders as a group button 1`] = `
<Button
  className="foo-class"
  context="danger"
  isBlock={false}
  isGroup={true}
  onBlur={[Function]}
  onClick={[Function]}
  size="sm"
  style={
    Object {
      "width": "1000px",
    }
  }
  text="This is text"
  type="button"
>
  <button
    className="gds-button foo-class gds-button--danger gds-button--sm gds-button-group__button"
    onBlur={[Function]}
    onClick={[Function]}
    style={
      Object {
        "width": "1000px",
      }
    }
    text="This is text"
    type="button"
  >
    My Cool Button
  </button>
</Button>
`;
